workers/gc/gcworker.py

```diff
@@ -39,13 +49,31 @@ def _garbage_collection_repos(self):
             assert features.GARBAGE_COLLECTION
 
             try:
-                repository = Repository.get(id=repo_ref.id)
-            except Repository.DoesNotExist:
-                return
+                with GlobalLock(
+                    "REPO_GARBAGE_COLLECTION_%s" % repo_ref.id, lock_ttl=REPOSITORY_GC_TIMEOUT + 60,
+                ) if not skip_lock_for_testing else empty_context():
+                    try:
+                        repository = Repository.get(id=repo_ref.id)
```

Comment:
Moving this `get by id` and the check for `marked_for_deletion` to outside the global lock will reduce the duration which a process is holding the lock (and thus contention). Whenever locking always move things outside the critical path.

Reply:
I actually put it into the lock to ensure that no one else is coming along and modifying before we check. This is the verification check (the lookup query also checks the state).

Reply:
It sounds like this check is unnecessary if the gc call also checks this value and is under the lock. The only value I see from checking the repository here is to avoid taking the lock before you find out you didn't need to gc.

Reply:
Yeah, I figure better to be careful, but it is extra code. Should I remove it?

Reply:
Yeah -- this is just wasting time under the lock. My vote is to totally remove it and only add it back before the lock is taken if we notice lock contention.